index.js

```diff
@@ -46,12 +47,15 @@ class Conf {
 			...options
 		};
 
+		const packageData = this._getPackageData();
```

Comment:
We may further simplify some more code by doing:

```js
               const packageData = this._getPackageData();

		options = {
			configName: 'config',
			fileExtension: 'json',
			projectSuffix: 'nodejs',
			clearInvalidConfig: true,
			serialize: value => JSON.stringify(value, null, '\t'),
			deserialize: JSON.parse,
			accessPropertiesByDotNotation: true,
                        projectName: packageData && packageData.name,
                        projectVersion: packageData && packageData.version,
			...options
		};
```
..and remove the assignment of these options from below.

Also note that having `this._getPackageData` is no longer necessary and we may inline it just like it was before. Just something to consider.

Reply:
Doesn't your suggestion always remove any specified `projectName` and `projectVersion` by the client?

At least for the name, it should only be inferred if the `cwd` option is not specified.

See https://github.com/sindresorhus/conf/commit/bb26c4f0a64cd8f1868eda2883cea651f0462fc0

Reply:
My suggestion just set the defaults of `options`, which the user may override. This means that if the client supplies a custom `projectName` and/or `projectVersion`, those will take precedence because we are mixin in.

Reply:
Checking https://github.com/sindresorhus/conf/commit/bb26c4f0a64cd8f1868eda2883cea651f0462fc0, the current code also re-introduces the bug because we are reading the package.json, even if we don't need to (both cwd and projectVersion were specified by the client).

To solve gracefully, ignore my comment above and perhaps do this instead:

```
    _fillOptionsFromPackageData(options) {
        // Avoid having to read package.json if not necessary
        // See: https://github.com/sindresorhus/conf/pull/62
        if (options.cwd && (!options.migrations || options.projectVersion)) {
              return;
        }

        const pkgPath = pkgUp.sync(parentDir);
	// Can't use `require` because of Webpack being annoying:
	// https://github.com/webpack/webpack/issues/196
        const pkgData = pkgPath && JSON.parse(fs.readFileSync(pkgPath, 'utf8'));

        if (!pkgData) {
             return;
        }

        if (!options.packageName) {
             options.packageName = pkgData.name;
        }

        if (!options.packageVersion) {
             options.packageVersion = pkgData.version;
        }
   }
```

and then use this function right after the `options` being assigned. You may then remove `_getPackageData` and all other unnecessary assignments that are already being done by this function.

Reply:
Another approach is to use something like this:

```js
// Place this inside the constructor 
const getPackageData = memoizeOne(() => {
        const pkgPath = pkgUp.sync(parentDir);
	// Can't use `require` because of Webpack being annoying:
	// https://github.com/webpack/webpack/issues/196
        const pkgData = pkgPath && JSON.parse(fs.readFileSync(pkgPath, 'utf8'));

       return pkgData || {};
});

// and then right below..

if (!options.cwd) {
   			if (!options.projectName) {
            			 options.projectName = getPackageData().name;
      			 }

        		// ...
}


if (migrations) {
   			if (!options.projectVersion) {
				options.projectVersion = getPackageData().version;
			}

			if (!options.projectVersion) {
				throw new Error('Project version could not be inferred. Please specify the `projectVersion` option.');
			}

			this._migrate(options.migrations, options.projectVersion);
		}
}
```

The advantage is that the code changes are minimal and most of the logic remains the same. `memoizeOne` refers to this: https://www.npmjs.com/package/memoize-one

Which one do you prefer?

Reply:
Spoiler: I prefer the last one, seems more simple and the options logic is more centralized as they are handled just in a single place and consistently.

Reply:
I also prefer the memoization solution for the reason that it scales if you need more properties from the package.json.

And also prefer the code flow of the second option where the project version is only set when needed (if the migrations are being used).

Reply:
Done, thank you very much for the nice suggestions 😄